From 08818f7dd94f46127e2ba7c8b5da43eca46c0f6e Mon Sep 17 00:00:00 2001
From: Christian Dreher <cdreher@prophesee.ai>
Date: Fri, 19 Jan 2024 12:46:13 +0100
Subject: [PATCH 10/13] [MV-1764] Use a DMABUF allocator when V4L2_HEAP is set

Each system defines its own heaps, there are several ways to chose the
right one:
- provide it through a config key, to be done in each app
- use a heap called metavision, and let the platform integrator create it
- let the user set it outside of Metavision

Here we use the third one, via an environment variable. It could have
been a config file.
---
 hal_psee_plugins/src/boards/v4l2/v4l2_board_command.cpp | 8 +++++++-
 1 file changed, 7 insertions(+), 1 deletion(-)

diff --git a/hal_psee_plugins/src/boards/v4l2/v4l2_board_command.cpp b/hal_psee_plugins/src/boards/v4l2/v4l2_board_command.cpp
index a50568ceff..5b86bcb988 100644
--- a/hal_psee_plugins/src/boards/v4l2/v4l2_board_command.cpp
+++ b/hal_psee_plugins/src/boards/v4l2/v4l2_board_command.cpp
@@ -152,7 +152,13 @@ std::unique_ptr<DataTransfer> V4L2BoardCommand::build_data_transfer(uint32_t raw
     // extract the /dev/videoX associated entity, and populate the DataTransfer with it.
     // Right now, we'll just hard code it to /dev/video0 ¯\_(ツ)_/¯
     // more details in: https://github.com/gjasny/v4l-utils/blob/master/utils/media-ctl/media-ctl.c#L526
-    return std::make_unique<V4l2DataTransfer>(device_->get_fd(), raw_event_size_bytes);
+
+    // If the environment set a heap, us it, otherwise, use the driver's allocator
+    if (std::getenv("V4L2_HEAP"))
+        return std::make_unique<V4l2DataTransfer>(device_->get_fd(), raw_event_size_bytes, "/dev/dma_heap",
+                                                  std::getenv("V4L2_HEAP"));
+    else
+        return std::make_unique<V4l2DataTransfer>(device_->get_fd(), raw_event_size_bytes);
 }
 
 } // namespace Metavision
-- 
2.39.2

